File: neetcode150/arrays_and_hashing/leetcode128.py
from typing import List

"""
Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence.

You must write an algorithm that runs in O(n) time.
"""


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        seen = set(nums)
        max_len = 0

        for num in nums:
            if num - 1 not in seen:
                i = num
                cnt = 0

                while i in seen:
                    cnt += 1
                    i += 1

                max_len = max(max_len, cnt)

        return max_len

File: js_algorithms/neetcode150/array_and_hashing/leetcode128.js
/**
 * 128. Longest Consecutive Sequence
 *
 * Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence.
 *
 * You must write an algorithm that runs in O(n) time.
 *
 * @param {number[]} nums
 * @return {number}
 */
const longestConsecutive = function (nums) {
  const uniqueNums = new Set(nums);
  let longestSeq = 0;

  for (const num of nums) {
    if (!uniqueNums.has(num - 1)) {
      let current = num;
      let currentSeq = 0;

      while (uniqueNums.has(current)) {
        currentSeq++;
        current++;
      }

      longestSeq = Math.max(longestSeq, currentSeq);
    }
  }

  return longestSeq;
};

module.exports = { longestConsecutive };
